File: apps/api/app/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from app.database import get_db
from app.models.category import Category, SubCategory
from app.models.program import Program
from app.models.pdf import PDF
from app.models.user import User
from app.auth import get_current_user, require_admin

router = APIRouter(prefix="/api", tags=["categories"])

def _cat_dict(c: Category) -> dict:
    return {
        "id": c.id,
        "programId": c.program_id,
        "categoryCode": c.category_code,
        "categoryName": c.category_name,
        "slug": c.slug,
        "categoryType": c.category_type,
        "description": c.description, "isActive": c.is_active, "icon": c.icon,
        "status": c.status,
        "isArchived": c.is_archived,
        "displayOrder": c.display_order,
        "created": c.created.isoformat() if c.created else None,
        "updated": c.updated.isoformat() if c.updated else None,
    }

def _subcat_dict(s: SubCategory) -> dict:
    return {
        "id": s.id, "subCategoryName": s.sub_category_name, "categoryId": s.category_id,
        "description": s.description, "isActive": s.is_active, "icon": s.icon,
        "displayOrder": s.display_order,
        "created": s.created.isoformat() if s.created else None,
        "updated": s.updated.isoformat() if s.updated else None,
    }


def _slugify(value: str) -> str:
    return "-".join(value.lower().strip().split())


async def _generate_category_code(
    db: AsyncSession,
    program_code: str,
    category_name: str,
) -> str:
    segment = "".join(ch for ch in category_name.upper() if ch.isalnum())[:4] or "GEN"
    prefix = f"{program_code}-{segment}"
    existing = await db.execute(
        select(Category.category_code).where(
            Category.category_code.is_not(None),
            Category.category_code.like(f"{prefix}-%"),
        )
    )
    max_seq = 0
    for row in existing.all():
        code = row[0] or ""
        parts = code.split("-")
        if len(parts) >= 3 and parts[-1].isdigit():
            max_seq = max(max_seq, int(parts[-1]))
    return f"{prefix}-{max_seq + 1:03d}"

# Categories
@router.get("/categories")
async def list_categories(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Category).order_by(Category.display_order))
    return {"items": [_cat_dict(c) for c in result.scalars().all()]}

@router.post("/categories")
async def create_category(body: dict, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    program_id = body.get("programId")
    category_name = body["categoryName"]
    slug = body.get("slug") or _slugify(category_name)

    program = None
    if program_id:
        program = await db.scalar(select(Program).where(Program.id == program_id))
        if not program:
            raise HTTPException(404, "Program not found")

    category_code = body.get("categoryCode")
    if not category_code and program:
        category_code = await _generate_category_code(db, program.program_code, category_name)

    cat = Category(
        program_id=program_id,
        category_code=category_code,
        category_name=category_name,
        slug=slug,
        category_type=body["categoryType"],
        description=body.get("description"),
        status=body.get("status", "active"),
        is_archived=body.get("isArchived", False),
        is_active=body.get("isActive", True),
        icon=body.get("icon"),
        display_order=body.get("displayOrder", 0),
    )
    db.add(cat)
    await db.commit()
    await db.refresh(cat)
    return _cat_dict(cat)

@router.patch("/categories/{cat_id}")
async def update_category(cat_id: str, body: dict, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    result = await db.execute(select(Category).where(Category.id == cat_id))
    cat = result.scalar_one_or_none()
    if not cat:
        raise HTTPException(404, "Not found")
    if "programId" in body:
        program = await db.scalar(select(Program).where(Program.id == body["programId"]))
        if not program:
            raise HTTPException(404, "Program not found")
        cat.program_id = program.id
    for k, v in {"categoryName": "category_name", "slug": "slug", "categoryType": "category_type",
                 "description": "description", "isActive": "is_active",
                 "status": "status", "isArchived": "is_archived",
                 "icon": "icon", "displayOrder": "display_order"}.items():
        if k in body:
            setattr(cat, v, body[k])
    await db.commit()
    await db.refresh(cat)
    return _cat_dict(cat)

@router.delete("/categories/{cat_id}")
async def delete_category(
    cat_id: str,
    force: bool = Query(False),
    archive: bool = Query(False),
    reassignTo: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_admin),
):
    result = await db.execute(select(Category).where(Category.id == cat_id))
    cat = result.scalar_one_or_none()
    if not cat:
        raise HTTPException(404, "Not found")

    pdf_count = await db.scalar(select(func.count()).select_from(PDF).where(PDF.category_id == cat_id))
    if pdf_count and pdf_count > 0:
        if reassignTo:
            replacement = await db.scalar(select(Category).where(Category.id == reassignTo))
            if not replacement:
                raise HTTPException(404, "Replacement category not found")
            if replacement.id == cat_id:
                raise HTTPException(400, "Cannot reassign to same category")
            pdfs = await db.execute(select(PDF).where(PDF.category_id == cat_id))
            for pdf in pdfs.scalars().all():
                pdf.category_id = replacement.id
        elif archive:
            cat.is_archived = True
            cat.status = "archived"
            cat.is_active = False
            await db.commit()
            await db.refresh(cat)
            return {"message": "Category archived", "item": _cat_dict(cat)}
        elif not force:
            raise HTTPException(
                409,
                "Category has PDFs. Reassign, archive, or use force delete.",
            )

    await db.delete(cat)
    await db.commit()
    return {"message": "Deleted"}

# SubCategories
@router.get("/subCategories")
async def list_subcategories(category_id: str = Query(None), db: AsyncSession = Depends(get_db)):
    q = select(SubCategory).order_by(SubCategory.display_order)
    if category_id:
        q = q.where(SubCategory.category_id == category_id)
    result = await db.execute(q)
    return {"items": [_subcat_dict(s) for s in result.scalars().all()]}

@router.post("/subCategories")
async def create_subcategory(body: dict, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    sub = SubCategory(
        sub_category_name=body["subCategoryName"],
        category_id=body["categoryId"],
        description=body.get("description"),
        is_active=body.get("isActive", True),
        icon=body.get("icon"),
        display_order=body.get("displayOrder", 0),
    )
    db.add(sub)
    await db.commit()
    await db.refresh(sub)
    return _subcat_dict(sub)

@router.patch("/subCategories/{sub_id}")
async def update_subcategory(sub_id: str, body: dict, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    result = await db.execute(select(SubCategory).where(SubCategory.id == sub_id))
    sub = result.scalar_one_or_none()
    if not sub:
        raise HTTPException(404, "Not found")
    for k, v in {"subCategoryName": "sub_category_name", "categoryId": "category_id",
                 "description": "description", "isActive": "is_active",
                 "icon": "icon", "displayOrder": "display_order"}.items():
        if k in body:
            setattr(sub, v, body[k])
    await db.commit()
    await db.refresh(sub)
    return _subcat_dict(sub)

@router.delete("/subCategories/{sub_id}")
async def delete_subcategory(sub_id: str, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    result = await db.execute(select(SubCategory).where(SubCategory.id == sub_id))
    sub = result.scalar_one_or_none()
    if not sub:
        raise HTTPException(404, "Not found")
    await db.delete(sub)
    await db.commit()
    return {"message": "Deleted"}

File: apps/api-node/src/routes/categories.ts
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { Category, Program, SubCategory, Subject } from "../models/index.js";
import { writeAudit } from "../lib/audit.js";
import { generateCategoryCode, slugify } from "../lib/codes.js";
import { listResponse, serializeDoc } from "../lib/serialize.js";
import { requireAuth, requirePermission } from "../plugins/auth.js";
import { PERMISSIONS } from "../lib/permissions.js";

const parseBody = (body: Record<string, unknown>) => ({
  program_id: (body.programId ?? body.program_id) as string | undefined,
  category_name: (body.categoryName ?? body.category_name) as string,
  category_type: (body.categoryType ?? body.category_type) as string,
  program_type: (body.programType ?? body.program_type) as string | undefined,
  category_code: (body.categoryCode ?? body.category_code) as string | undefined,
  slug: body.slug as string | undefined,
  description: body.description as string | undefined,
  is_active: (body.isActive ?? body.is_active) as boolean | undefined,
  display_order: (body.displayOrder ?? body.display_order) as number | undefined
});

export const registerCategoryRoutes = async (app: FastifyInstance): Promise<void> => {
  app.get("/api/categories", { preHandler: requireAuth }, async () => {
    const rows = await Category.find({ is_archived: { $ne: true } }).sort({ display_order: 1, created: -1 }).lean();
    const { Pdf, SubCategory: SubCat } = await import("../models/index.js");
    const [counts, subCats] = await Promise.all([
      Pdf.aggregate([
        { $match: { deleted_at: null } },
        { $group: { _id: "$category_id", count: { $sum: 1 } } }
      ]),
      SubCat.find({}).sort({ display_order: 1 }).lean()
    ]);
    const countMap = new Map(counts.map((c) => [c._id, c.count]));
    const subCatMap = new Map<string, typeof subCats>();
    for (const s of subCats) {
      const arr = subCatMap.get(s.category_id) ?? [];
      arr.push(s);
      subCatMap.set(s.category_id, arr);
    }
    const items = rows.map((r) => ({
      ...serializeDoc(r as Record<string, unknown>),
      pdfCount: countMap.get(r.id) ?? 0,
      subCategories: (subCatMap.get(r.id) ?? []).map((s) => serializeDoc(s as Record<string, unknown>))
    }));
    return listResponse(items);
  });

  // Get all PDFs mapped to a specific category
  app.get("/api/categories/:cat_id/pdfs", { preHandler: requireAuth }, async (request, reply) => {
    const params = z.object({ cat_id: z.string() }).parse(request.params);
    const query = z
      .object({
        page: z.coerce.number().default(1),
        per_page: z.coerce.number().default(50),
        status: z.string().optional(),
        is_active: z.coerce.boolean().optional()
      })
      .parse(request.query);

    const cat = await Category.findOne({ id: params.cat_id });
    if (!cat) return reply.status(404).send({ detail: "Category not found" });

    const { Pdf, SubCategory: SubCat } = await import("../models/index.js");
    const { enrichPdfs } = await import("../lib/pdfEnrich.js");

    const filter: Record<string, unknown> = { category_id: params.cat_id, deleted_at: null };
    if (query.status) filter.status = query.status;
    if (query.is_active !== undefined) filter.is_active = query.is_active;

    const skip = (query.page - 1) * query.per_page;
    const [rows, total] = await Promise.all([
      Pdf.find(filter).sort({ created: -1 }).skip(skip).limit(query.per_page).lean(),
      Pdf.countDocuments(filter)
    ]);

    const enriched = await enrichPdfs(rows as Record<string, unknown>[]);
    return listResponse(enriched, total);
  });

  app.post(
    "/api/categories",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request, reply) => {
      const raw = parseBody(z.record(z.string(), z.unknown()).parse(request.body));
      if (!raw.category_name) {
        return reply.status(400).send({ detail: "categoryName is required" });
      }
      raw.category_type = raw.category_type ?? "general";
      let category_code = raw.category_code;
      const slug = raw.slug ?? slugify(raw.category_name);
      if (raw.program_id) {
        const program = await Program.findOne({ id: raw.program_id });
        if (!program) return reply.status(404).send({ detail: "Program not found" });
        if (!category_code) {
          category_code = await generateCategoryCode(program.program_code, raw.category_name);
        }
      }
      const cat = await Category.create({
        program_id: raw.program_id ?? null,
        category_code,
        category_name: raw.category_name,
        slug,
        category_type: raw.category_type,
        program_type: raw.program_type ?? "pdf",
        description: raw.description,
        is_active: raw.is_active ?? true
      });
      if (request.authUser?.sub) {
        await writeAudit({
          user_id: request.authUser.sub,
          action: "category_create",
          resource_type: "category",
          resource_id: cat.id,
          request
        });
      }
      return serializeDoc(cat.toObject() as Record<string, unknown>);
    }
  );

  app.patch(
    "/api/categories/:cat_id",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request, reply) => {
      const params = z.object({ cat_id: z.string() }).parse(request.params);
      const body = z.record(z.string(), z.unknown()).parse(request.body);
      const update: Record<string, unknown> = {};
      if (body.categoryName !== undefined) update.category_name = body.categoryName;
      if (body.category_name !== undefined) update.category_name = body.category_name;
      if (body.categoryType !== undefined) update.category_type = body.categoryType;
      if (body.programType !== undefined) update.program_type = body.programType;
      if (body.programId !== undefined) update.program_id = body.programId;
      if (body.isActive !== undefined) update.is_active = body.isActive;
      if (body.isArchived !== undefined) update.is_archived = body.isArchived;
      if (body.displayOrder !== undefined) update.display_order = body.displayOrder;
      if (body.description !== undefined) update.description = body.description;
      const updated = await Category.findOneAndUpdate({ id: params.cat_id }, { $set: update }, { new: true });
      if (!updated) return reply.status(404).send({ detail: "Category not found" });
      return serializeDoc(updated.toObject() as Record<string, unknown>);
    }
  );

  app.delete(
    "/api/categories/:cat_id",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request, reply) => {
      const params = z.object({ cat_id: z.string() }).parse(request.params);
      const childCount = await SubCategory.countDocuments({ category_id: params.cat_id });
      if (childCount > 0) {
        return reply.status(409).send({ detail: `Cannot delete: ${childCount} class(es) exist under this program. Delete or move them first.` });
      }
      const deleted = await Category.findOneAndDelete({ id: params.cat_id });
      if (!deleted) return reply.status(404).send({ detail: "Category not found" });
      return { message: "Category deleted" };
    }
  );

  app.get("/api/subCategories", { preHandler: requireAuth }, async (request) => {
    const query = z
      .object({
        category_id: z.string().optional(),
        categoryId: z.string().optional(),
        is_active: z.coerce.boolean().optional()
      })
      .parse(request.query);
    const categoryId = query.category_id ?? query.categoryId;
    const filter: Record<string, unknown> = categoryId ? { category_id: categoryId } : {};
    if (query.is_active !== undefined) filter.is_active = query.is_active;
    const rows = await SubCategory.find(filter).sort({ display_order: 1 }).lean();
    return listResponse(rows.map((r) => serializeDoc(r as Record<string, unknown>)));
  });

  app.post(
    "/api/subCategories",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request) => {
      const body = z.record(z.string(), z.unknown()).parse(request.body);
      const sub = await SubCategory.create({
        sub_category_name: (body.subCategoryName ?? body.sub_category_name) as string,
        category_id: (body.categoryId ?? body.category_id) as string,
        sub_category_code: (body.subCategoryCode ?? body.sub_category_code) as string | undefined,
        description: body.description as string | undefined,
        is_active: (body.isActive ?? body.is_active ?? true) as boolean,
        display_order: (body.displayOrder ?? body.display_order ?? 0) as number
      });
      return serializeDoc(sub.toObject() as Record<string, unknown>);
    }
  );

  app.patch(
    "/api/subCategories/:sub_id",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request, reply) => {
      const params = z.object({ sub_id: z.string() }).parse(request.params);
      const body = z.record(z.string(), z.unknown()).parse(request.body);
      const update: Record<string, unknown> = {};
      if (body.subCategoryName !== undefined) update.sub_category_name = body.subCategoryName;
      if (body.categoryId !== undefined) update.category_id = body.categoryId;
      if (body.isActive !== undefined) update.is_active = body.isActive;
      if (body.subCategoryCode !== undefined) update.sub_category_code = body.subCategoryCode;
      if (body.description !== undefined) update.description = body.description;
      if (body.displayOrder !== undefined) update.display_order = body.displayOrder;
      const updated = await SubCategory.findOneAndUpdate({ id: params.sub_id }, { $set: update }, { new: true });
      if (!updated) return reply.status(404).send({ detail: "Sub-category not found" });
      return serializeDoc(updated.toObject() as Record<string, unknown>);
    }
  );

  app.delete(
    "/api/subCategories/:sub_id",
    { preHandler: requirePermission(PERMISSIONS.CATEGORY_MANAGE) },
    async (request, reply) => {
      const params = z.object({ sub_id: z.string() }).parse(request.params);
      const childCount = await Subject.countDocuments({ class_id: params.sub_id });
      if (childCount > 0) {
        return reply.status(409).send({ detail: `Cannot delete: ${childCount} subject(s) exist under this class. Delete them first.` });
      }
      const deleted = await SubCategory.findOneAndDelete({ id: params.sub_id });
      if (!deleted) return reply.status(404).send({ detail: "Sub-category not found" });
      return { message: "Sub-category deleted" };
    }
  );
};
